### PR TITLE
Add more trait implementations for [T; $N]

### PR DESCRIPTION
**Pull request by SimonSapin**
A number of traits are implemented for `[T; $N]` in a macro with `$N` between 0 and 32.

Ideally, the language would support a generic integer `N` in a single `impl`. In the meantime, the macro is what we have.

This adds more values on `$N` in the macro: powers of 2 and 10 up to 2^30. This doubles the total number of `impl`s generated by this macro.

I’m experimenting with a `String`-like type whose backing storage is generic on `T: AsMut<[u8]>`. Buffers like `[u8; 16 * 1024]` are reasonable to allocate on the stack (`std::io::copy` does it) but would previously not implement `AsMut<[u8]>`.

### CODE REVIEW

File: src/libcore/array.rs [hunk 1]
```diff
@@ -186,6 +186,29 @@ array_impls! {
...
+    64 128 256 512 1024
+    2 * 1024
+    4 * 1024
+    8 * 1024
+    16 * 1024
+    32 * 1024
+    64 * 1024
+    128 * 1024
+    256 * 1024
+    515 * 1024
```
Comment (petrochenkov) [inline src/libcore/array.rs]: 515 -> 512

File: src/libcore/array.rs [hunk 2]
```diff
@@ -186,6 +186,29 @@ array_impls! {
...
+    1024 * 1024
+    2 * 1024 * 1024
+    4 * 1024 * 1024
+    8 * 1024 * 1024
+    16 * 1024 * 1024
+    32 * 1024 * 1024
+    64 * 1024 * 1024
+    128 * 1024 * 1024
+    256 * 1024 * 1024
+    515 * 1024 * 1024
```
Comment (petrochenkov) [inline src/libcore/array.rs]: And here as well

File: src/libcore/array.rs (hunk 1 above)
Comment (SimonSapin) [inline src/libcore/array.rs]: Good catch! Fixed.

**Comment by Gankro** [comment]
Didn't we have to back off on more of these impls because it makes the compiler deeply unhappy? Can't remember the details. Possibly that was a bunch of `<T, &T>, <&T, T>, <T, T> ...` impls?

**Comment by SimonSapin** [comment]
I don’t know, other than libcoretest compiles and passes with this change.

**Comment by arielb1** [comment]
@alexcrichton 

A good amount of the metadata bloat was due to a few issues I fixed. Could we measure it now?

**Comment by alexcrichton** [comment]
Yeah if we've gotten the file size under control then these wouldn't be too bad to add I don't think.

**Comment by alexcrichton** [comment]
@SimonSapin, can you measure the impact of these impls on the size of the libcore artifact?

**Comment by SimonSapin** [comment]
Do you mean the size of `x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-bb943c5a.rlib` before and after this change? Ok, I’ll do it later. (Rebuilding takes a while.)

**Comment by SimonSapin** [comment]
On commit 05eb81ae11929438f2386d34113e45a5a2c3552b (which this PR branches from), `x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-bb943c5a.rlib` is 16,277,278 bytes. With this PR, it’s 18,351,834 bytes, 12% bigger.

**Comment by alexcrichton** [comment]
I personally feel that tacking on 2MB of extra metadata for this sort of data isn't worth it right now, but I'll bring this up in the next triage meeting to see what others think (although discussion can certainly happen before that!)

**Comment by SimonSapin** [comment]
I’m a bit surprised this is this much data, but oh well. (I’m not so attached to this PR as I ended up making [a custom trait](https://github.com/SimonSapin/rust-std-candidates/blob/fa4ff4da0e513088e5c606515c63c7e78df75644/string-wrapper/lib.rs#L17-L22) anyway.)

**Comment by alexcrichton** [comment]
The libs team discussed this during triage today and the decision was to close this for now. We're fine adding impls for a fixed number of arrays (e.g. the 0-32 we already have), but adding more numbers like this moves into the "a little bit too arbitrary" territory that we'd prefer to hold off for a more proper generic mechanism.

Thanks regardless for the PR though @SimonSapin!

**Comment by ryankurte** [comment]
+1 to the wish for either a larger arbitrary range (at least 0-64) and/or inclusion of common CS and cryptographic primitive sizes (256, 384, 512, 1024, 2048, 4096). seems like both would be an easy solution to a common usability problem until const generics land.

**Pull request closed without merging**